File: 1405-longest-happy-string/1405-longest-happy-string.c
#include <stdio.h>

/**
 * @brief Generates the longest happy string using the given counts of 'a', 'b', and 'c'.
 *
 * @param a (int): Number of 'a' characters.
 * @param b (int): Number of 'b' characters.
 * @param c (int): Number of 'c' characters.
 *
 * @return (char*): The longest happy string.
 */

char* longestDiverseString(int a, int b, int c) {
    static char result[201];  // Maximum length is 100 a's + 100 b's + 100 c's = 300 + 1 for '\0'
    int index = 0;
    
    // Run the loop while there are characters left to be used
    while (a > 0 || b > 0 || c > 0) {
        // Check the current result to avoid 3 consecutive same characters
        if ((a >= b && a >= c) && !(index > 1 && result[index - 1] == 'a' && result[index - 2] == 'a')) {
            result[index++] = 'a';  // Add 'a' to the result
            a--;
        } else if ((b >= a && b >= c) && !(index > 1 && result[index - 1] == 'b' && result[index - 2] == 'b')) {
            result[index++] = 'b';  // Add 'b' to the result
            b--;
        } else if (c > 0 && !(index > 1 && result[index - 1] == 'c' && result[index - 2] == 'c')) {
            result[index++] = 'c';  // Add 'c' to the result
            c--;
        } else {
            // If we are here, it means we have two consecutive characters and need to switch
            // Check if other characters are available and append one of them.
            if (a > 0 && !(index > 1 && result[index - 1] == 'a')) {
                result[index++] = 'a';
                a--;
            } else if (b > 0 && !(index > 1 && result[index - 1] == 'b')) {
                result[index++] = 'b';
                b--;
            } else if (c > 0 && !(index > 1 && result[index - 1] == 'c')) {
                result[index++] = 'c';
                c--;
            } else {
                // If no valid characters are available, break out of the loop
                break;
            }
        }
    }
    
    result[index] = '\0';  // Null-terminate the string
    return result;
}

File: Daily Questions/1405-longest-happy-string/1405-longest-happy-string.py
from heapq import heappush, heappop

class Solution:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        """
        Generate the longest happy string containing at most a 'a's, b 'b's, and c 'c's.
        
        Parameters:
        a (int): Number of 'a' characters available.
        b (int): Number of 'b' characters available.
        c (int): Number of 'c' characters available.
        
        Returns:
        str: The longest happy string that can be generated.
        """
        # Max heap to always pick the character with the most count left
        heap = []
        
        # Push negative values because Python's heapq is a min-heap by default
        if a > 0:
            heappush(heap, (-a, 'a'))
        if b > 0:
            heappush(heap, (-b, 'b'))
        if c > 0:
            heappush(heap, (-c, 'c'))
        
        result = []
        
        while heap:
            # Get the most frequent character
            count1, char1 = heappop(heap)
            
            # If the last two characters in result are the same as char1, we need to use another character
            if len(result) > 1 and result[-1] == result[-2] == char1:
                if not heap:
                    break  # No other character is available, exit
                count2, char2 = heappop(heap)
                result.append(char2)
                if count2 + 1 < 0:
                    heappush(heap, (count2 + 1, char2))  # Add it back with one less count
                heappush(heap, (count1, char1))  # Push the first character back into the heap
            else:
                # Append the current character to the result
                result.append(char1)
                if count1 + 1 < 0:
                    heappush(heap, (count1 + 1, char1))  # Add it back with one less count
        
        return ''.join(result)